File: Ruby/Civitas/lib/sorpresa_salir_carcel.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

module Civitas
  class SorpresaSalirCarcel < Sorpresa
    def initialize(mazo)
      super()
      @texto="Carta para evitar la cárcel. \n"
      @mazo=mazo
  
      #puts "Sorpresa SalirCarcel creada."
    end
    
    # Método sobrescrito
    def aplicar_a_jugador(actual, todos)
      if(jugador_correcto(actual, todos))
        informe(actual, todos)
        
        no_salvoc=true
        
        for i in (0..todos.length)
          if(todos[i].tiene_salvoconducto)
            no_salvoc=false
          end
        end
        
        if(no_salvoc)
          todos[actual].obtener_salvoconducto(self)
          salir_del_mazo
        end
      end
    end
    
    def salir_del_mazo
        self.mazo.habilitar_carta_especial(self)
    end
  
    def usada
        self.mazo.habilitar_carta_especial(self)
    end
       
    def to_s
      "\n\t ¡SORPRESA! \n
       \n Has recibido: 
       \n #{@texto} \n"
    end
  end
end

File: Java/Civitas/src/civitas/SorpresaSalirCarcel.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

import java.util.ArrayList;

/**
 *
 * @author arturo
 */
public class SorpresaSalirCarcel extends Sorpresa{
    private mazoSorpresas mazo;
    
    SorpresaSalirCarcel(mazoSorpresas mazo, String texto){
        super(texto);
        this.mazo=mazo;
    }
    
    @Override
    public void aplicarAJugador(int actual, ArrayList<Jugador> todos){
        if(jugadorCorrecto(actual, todos)==true){
        informe(actual, todos);
             boolean salvoconducto=false;
            for(int i=0; i<3; i++){
                if(todos.get(i).tieneSalvoconducto()==true){
                    salvoconducto=true;
                }
            }
            if(salvoconducto==false){
                todos.get(actual).obtenerSalvoconducto(this);
                this.salirDelMazo();
            }
        }
    }
    
    void salirDelMazo(){
        mazo.inhabilitarCartaEspecial(this);
    }
    
    void Usada(){
        mazo.habilitarCartaEspecial(this);
    }
    
    @Override
    public String toString(){
        return "Sorpresa: "+super.getTexto()+"\nSales de la carcel\n";
    }
}
